lib/SILOptimizer/IPO/ClosureSpecializer.cpp

```diff
@@ -657,11 +640,24 @@ ClosureSpecCloner::initCloned(SILOptFunctionBuilder &FunctionBuilder,
                       ? ParameterConvention::Direct_Unowned
                       : ParameterConvention::Direct_Owned;
     }
-
-    SILParameterInfo NewPInfo(PInfo.getInterfaceType(), ParamConv);
+    auto newParamType =
+        PInfo.getInterfaceType(); // .subst(partialApply->getSubstitutionMap());
+    SILParameterInfo NewPInfo(newParamType, ParamConv);
     NewParameterInfoList.push_back(NewPInfo);
   }
 
+  SmallVector<SILResultInfo, 2> results;
+  for (auto result : ClosureUserFunTy->getResults()) {
+    auto newTy = result.getInterfaceType().subst(
+        CallSiteDesc.getApplyCallee()->getForwardingSubstitutionMap());
+    if (newTy->hasArchetype()) {
```

Comment:
What's going on here?

Reply:
We're re-building the parameters in for the new function that will be cloned. This essentially the same operation as what we do for the other parameters (these are just results, though). The logic is, if there's an archetype, just add that as the result, we're not going to get any better and the cloned function will have the same generic environment. Otherwise, if we've got an "actual" type, we can just use that as the result.